### ISSUE TITLE
bug in \iftranslationdefined

### ISSUE THREAD

**Issue by josephwright** [josephwright/beamer#84]
The macros `\newtranslation`, `\renewtranslation` and `\providetranslation` seem not to work correctly due to a bug in `\iftranslationdefined`. A patch that should fix this issue is attached.
````
--- translator.sty~
+++ translator.sty
@@ -34,12 +34,12 @@
 \newcommand\iftranslationdefined[4][]{%
   \def\trans@to{\trans@defaultto}%
   \setkeys{translator}{#1}%
-  \expandafter\ifx\csname tr@@@\trans@to @#2\endcsname\@undefined%
+  \expandafter\ifx\csname tr@@@\trans@to @#2\endcsname\relax%
     \let\trans@next=\@firstoftwo%
   \else%
     \let\trans@next=\@secondoftwo%
   \fi%
-  \trans@next{#3}{#4}%
+  \trans@next{#4}{#3}%
 }
 
 \newcommand\deftranslation[3][]{%
````

----------------------------------------
- Bitbucket: https://bitbucket.org/rivanvx/beamer/issue/83
- Originally reported by: mhp
- Originally created at: 2010-10-04

**Comment by josephwright** [comment]
_From Vedran Miletić on 2010-10-26_

Can you explain in what way they don't work correctly?

**Comment by josephwright** [comment]
_From Anonymous on 2010-10-31_

Consider the following minimal example:
````
\documentclass[ngerman, english]{article}
\usepackage{babel}
\usepackage{translator}
\newtranslation[to=German]{Term}{Begriff}
\renewtranslation[to=German]{Term}{Ausdruck}
\begin{document}
\selectlanguage{ngerman}
\translate{Term}
\end{document}
````
`\renewtranslation` issues the error:
````
! Package translator Error: Translation of ``Term'' not defined.. 
````
Clearly, the translation of `Term` has been defined before.

The reason for this lies in the definition of `\iftranslationdefined`:
````
\newcommand\iftranslationdefined[4][]{%
  ... \expandafter\ifx\csname tr@@@\trans@to @#2\endcsname\undefined%
  \let\trans@next=\@firstoftwo%
\else%
\let\trans@next=\@secondoftwo%
\fi%
\trans@next{#3}{#4}%
}

Assuming that the meaning of `\undefined` is undefined (and not `\relax`) the condition `\expandafter\ifx\csname tr@@@\trans@to @#2\endcsname\undefined` is never true since `\csname ... \endcsname` gives the meaning of \relax to any undefined control sequence, i.e. the first branch is never interpreted. Moreover, the arguments passed to `\trans@next` have to be exchanged.

**Comment by josephwright** [comment]
_From mhp on 2012-11-28_

The bug in `\iftranslationdefined` persists. Thus, processing the MWE
````
\documentclass[ngerman, english]{article}

\usepackage{babel}
\usepackage{translator}

\newtranslation[to=German]{Term}{Begriff}
\renewtranslation[to=German]{Term}{Ausdruck}

\begin{document}

\selectlanguage{ngerman}

\translate{Term}

\end{document}
````
still leads to `\renewtranslation` issuing the error message
````
! Package translator Error: Translation of ``Term'' not defined..
````
The following patch seems to fix this bug:
````
\usepackage{xpatch}

\makeatletter

\xpatchcmd{\iftranslationdefined}{%
  \ifx\csname tr@@@\trans@to @#2\endcsname\@undefined
    \let\trans@next=\@firstoftwo
  \else
    \let\trans@next=\@secondoftwo
  \fi
}{%
  \ifx\csname tr@@@\trans@to @#2\endcsname\relax
    \let\trans@next=\@secondoftwo
  \else
    \let\trans@next=\@firstoftwo
  \fi
}{
}{
}

\makeatother
````

**Comment by josephwright** [comment]
Fixed by 265b70843ba59210139da8d9034ffcab97edfe45
